Feature: Local State

fsaravia opened the pull request:
This PR introduces a `Shield` that encapsulates the variables used to configure `shield`. It also adds a convenience `shield.Default()` function that returns the struct initialized with what used to be the default values for the constants.

There are two main motivations for this:
- Using constants goes against flexibility in the sense that a user cannot alter those values anyhow.
- Moving everything inside a struct guarantees that a user can have multiple instances of `shield`, each one with their own config, without issues.

fsaravia committed:
Use variables to configure shield and encapsulate them in a struct

fsaravia committed:
Add shield.Default() to return an instance with default values

fsaravia committed:
Use a different instance of Shield on each test run

fsaravia committed:
Add documentation for Shield and Default()

fsaravia committed:
Remove unused constants

fsaravia committed:
Make newShield a variable of type func

fsaravia commented:
@inkel I made this based on our chat today. Let me know your thoughts :)

inkel commented on shield.go:
```diff
@@ -9,27 +9,41 @@ import (
+//     maxLength: indicates the maximum length that a password can have.
+//     saltLength: is the length of the salt generated by GenerateSalt.
+type Shield struct {
+	iterations int
```
Naming fields like this result in non-exported fields, thus the user can never explicitly set those values, it only works by passing the values in a positional way. Also the following is shorter :wink: 

```go
type Shield struct {
    Iterations, MaxLength, SaltLength int
}
```

inkel commented on shield.go:
```diff
@@ -9,27 +9,41 @@ import (
+//     saltLength: is the length of the salt generated by GenerateSalt.
+type Shield struct {
+	iterations int
+	maxLength  int
```
I'm not sure this should be a configurable value, but a constant instead. `4096` was chosen as a constant because, as it [says in `Shield::Password` comments](https://github.com/cyx/shield/blob/master/lib/shield.rb#L103-L109): a very long password would hang a server running this operation, thus DoS'ing it. Check the [referenced blog post](https://www.djangoproject.com/weblog/2013/sep/15/security/) as well.

inkel commented on shield.go:
```diff
@@ -9,27 +9,41 @@ import (
+		iterations: 5000,
+		maxLength:  4096,
+		saltLength: 32}
+}
```
I wouldn't define this as a function, but I'd do what other packages in the stdlib do:

```go
var DefaultShield = Shield{Iterations: 5000, SaltLength: 32}
```

inkel replied:
Maybe `Default` is a better name than `DefaultShield`.

inkel replied:
Not sure I follow :smile_cat:

fsaravia replied:
Nevermind, I like your solution the better :)

inkel commented on shield_test.go:
```diff
@@ -6,12 +6,14 @@ import (
 	"testing"
 )
 
+var newShield = Default
```
Why not assigning the return value of `Default` instead? Or better yet, using the `DefaultShield` variable? This will also save you of having to call `newShield()` on each test function.

inkel commented on shield.go:
```diff
@@ -38,17 +52,17 @@ func digest(password, salt []byte) string {
 // random bytes.
-func GenerateSalt() ([]byte, error) {
-	salt := make([]byte, SaltLength)
+func (s Shield) GenerateSalt() ([]byte, error) {
```
There's no need to `GenerateSalt` to be a _method_ of `Shield` I think. Let's discuss it.

inkel commented:
@fsaravia thank  you! I've left a couple of comments. One thing that I mention to tell you in our chat yesterday is that I was envisioning something like this:

```go
package main

import "github.com/inkel/shield"

func check(password string, salt []byte, s shield.Shield) {
  crypted, _  := shield.Encrypt(password, salt)
  if !shield.Check(password, crypted) {
    panic(":boom:")
  }
}

func main() {
  password := "lorem"
  salt, _ := shield.GenerateSalt()
  
  check(password, salt, shield.DefaultShield)
  
  // shield.Shield{Iterations: 5000} looks a bit odd, but maybe it's better than my idea
  s := shield.New(shield.Options{ Iterations: 5000 })

  check(password, salt, s)
}
```

And just for easiness for using the library, I'd also have the following public functions:

```go
func Encrypt(password string, salt []byte) (string, error)
  return DefaultShield.Encrypt(password, salt)
}

func Check(password, crypted string) bool {
  return DefaultShield.Check(password, crypted)
}
```

What do you think?

(Yes, I've changed the method signatures. I haven't decided on that yet.)

fsaravia committed:
Make Shield variables exported

fsaravia committed:
Keep MaxLength as a constant instead of a variable

fsaravia committed:
GenerateSalt doesn't need to be a method

fsaravia committed:
Avoid indirection on the Default function

fsaravia commented:
@inkel I addressed your feedback except for the one I answered. I also have some more ideas about how to improve this library but I feel they are out of scope of this PR. Let me know your thoughts and we can either merge this one and continue working on a different branch or close this one and use a different approach, your call.

inkel commented on shield.go:
```diff
@@ -8,31 +8,45 @@ import (
+func Default() Shield {
+	return Shield{
+		Iterations: 5000,
+		SaltLength: 32}
```
`s/}/,\n}/` :wink:

fsaravia committed:
Use a variable for the default instance of Shield
